File: lab4/Entities/UserMonth.cs
using lab4.Helpers;

namespace lab4.Entities
{
    public class UserMonthJson : EntityJson
    {
        public DateTime month { get; set; } = new DateTime();
        public string userName { get; set; } = "";
        public bool frozen { get; set; } = false;
        public ICollection<UserActivityJson>? userActivities { get; set; }
        public ICollection<ApprovedActivityJson>? approvedActivities { get; set; }
    }

    public class UserMonthT
    {
        public string? User { get; set; }
        public int? Time {get; set; }
    }

    public class UserMonth : IEntity<UserMonthJson>
    {
        // PK
        public DateTime Month { get; set; } = new DateTime();
        public string UserName { get; set; } = "";

        public bool Frozen { get; set; } = false;

        // Parents
        public User? User { get; set; }

        // Children
        public virtual ICollection<UserActivity>? UserActivities { get; set; }
        public virtual ICollection<ApprovedActivity>? ApprovedActivities { get; set; }

        public UserMonth()
        {
            this.UserActivities = new List<UserActivity>();
            this.ApprovedActivities = new List<ApprovedActivity>();
        }

        public UserMonthJson toJSON()
        {
            return new UserMonthJson
            {
                month=Month,
                userName=UserName,
                frozen=Frozen,
                userActivities= Helper.toJSON<UserActivity, UserActivityJson>(this.UserActivities),
                approvedActivities= Helper.toJSON<ApprovedActivity, ApprovedActivityJson>(this.ApprovedActivities),
            };
        }
    }
}

File: lab4/ClientApp/src/entities/UserMonth.ts
import { gremoveFromArray } from "../Helpers";
import ApprovedActivity from "./ApprovedActivityStats";
import UserActivity from "./UserActivity";

export class UserMonth {
    month: Date
    userName: string;
    frozen: boolean;
    userActivities: UserActivity[];
    approvedActivities: ApprovedActivity[];

    constructor(month: Date, userName: string, frozen: boolean, userActivities: UserActivity[], approvedActivities: ApprovedActivity[]) {
        this.month = month;
        this.userName = userName;
        this.frozen = frozen;
        this.userActivities = userActivities;
        this.approvedActivities = approvedActivities;
    }

    toJSON() {
        return {
            month: this.month,
            userName: this.userName,
            frozen: this.frozen,
            userActivities: this.userActivities,
            approvedActivities: this.approvedActivities
        };
    }

    static fromJSON(json: any) {
        var uas = json['userActivities'].map((ua: any) => UserActivity.fromJSON(ua));
        return new UserMonth(json['month'], json['userName'], json['frozen'], uas, json['approvedActivities']);
    }

    static createEmpty() {
        return new UserMonth(new Date(), '', false, [], []);
    }

    modify(setter: Function, key: string, value: any){
        setter(UserMonth.fromJSON({...this.toJSON(), [key]:value}));
    }
};

export default UserMonth;
